Update the core commands to match the latest IRI 1.8.1 release

pdecol opened the pull request:
resolves #235

pdecol commented:
I usually like it to clean up code that I touch. Since making the code PEP8 conform is still work in progress I changed the indentation level to 4 spaces in existing test files where I added more tests. That is the reason why this pull request has quite a lot of line changes. If its prefered to have these changes in a separate ticket/pull request let me know and I can change it back to 2 spaces.

pdecol commented on iota/commands/core/get_balances.py:
```diff
@@ -44,19 +45,50 @@ def __init__(self):
             'balances': f.Array | f.FilterRepeater(f.Int),
 
-            'milestone':
-                f.ByteString(encoding='ascii') | Trytes(Address),
```
Note that this is a potential backwards incompatible change. `milestone` was removed in IRI version 1.4.1.6 (https://github.com/iotaledger/iri/blob/v1.4.1.6/src/main/java/com/iota/iri/service/dto/GetBalancesResponse.java) which was released on the 6th of January 2018. Since this was more than one and a half years ago I think it is justified to remove it here too.

todofixthis replied:
👍🏻 Sounds good to me.

todofixthis commented:
Hi @pdecol thanks for working on this!  Really appreciate the attention to detail and being proactive about keeping the codebase consistent!

Will have a look at this today 👍🏻

lzpap commented on iota/commands/core/get_missing_transactions.py:
```diff
@@ -0,0 +1,47 @@
+class GetMissingTransactionsResponseFilter(ResponseFilter):
+    def __init__(self):
+        super(GetMissingTransactionsResponseFilter, self).__init__({
+            'hashes':
```
hey, I have a question about response filters. Why is the duration field never filtered in responses?

todofixthis replied:
Good question.  The approach I went for with these was to be restrictive towards requests (because those are within the developer's control) and permissive towards responses (because those are up to the IRI node and out of the developer's control).

Response filters generally only do basic conversions (e.g., for addresses, transaction hashes, etc.); I try to avoid situations where a response from the node would generate an exception in the response filter (principle of least astonishment).

todofixthis replied:
That said, I can see the other side of the argument — if the response from the node contains missing/unexpected values, that would be more astonishing than if the lib raised an exception \[that could be caught and handled accordingly in the app\].

For the same reason, it's never really sat well with me that `get_bundles()` doesn't make the validation optional (done this way to match the behaviour of the JS lib at the time, but my preference would be to disable validation by default unless the caller explicitly requests it).

I've never made any of this explicit in the code, so might be good to make a decision one way or the other, and then document it somewhere.

lzpap commented on test/commands/core/get_node_api_configuration_test.py:
```diff
@@ -0,0 +1,95 @@
+        )
+
+
+class GetNodeAPIConfigurationFilterTestCase(BaseFilterTestCase):
```
```suggestion
class GetNodeAPIConfigurationResponseFilterTestCase(BaseFilterTestCase):
```

todofixthis commented on iota/commands/core/check_consistency.py:
```diff
@@ -36,3 +36,10 @@ def __init__(self):
+    def __init__(self):
+        super(CheckConsistencyResponseFilter, self).__init__({
+            'state': f.Required | f.Type(bool),
+        })
```
❔ Pending https://github.com/iotaledger/iota.lib.py/pull/236/files#r327037008

todofixthis commented on iota/commands/core/get_node_api_configuration.py:
```diff
@@ -0,0 +1,45 @@
+            'maxGetTrytes': f.Type(int),
+            'maxBodyLength': f.Type(int),
+            'testNet': f.Type(bool),
+            'milestoneStartIndex': f.Type(int),
```
❔ Pending https://github.com/iotaledger/iota.lib.py/pull/236/files#r327037008

todofixthis commented on iota/commands/core/were_addresses_spent_from.py:
```diff
@@ -37,3 +37,14 @@ def __init__(self):
+                    f.Required |
+                    f.Type(bool),
+                ),
+        })
```
❔ Pending https://github.com/iotaledger/iota.lib.py/pull/236/files#r327037008

todofixthis commented on test/commands/core/get_missing_transactions_test.py:
```diff
@@ -0,0 +1,116 @@
+from iota.commands.core import GetMissingTransactionsCommand
+
+
+class GetMissingTransactionsFilterTestCase(BaseFilterTestCase):
```
```suggestion
class GetMissingTransactionsRequestFilterTestCase(BaseFilterTestCase):
```

todofixthis commented on test/commands/core/get_missing_transactions_test.py:
```diff
@@ -0,0 +1,116 @@
+
+    def test_pass_empty(self):
+        """
+        The incoming response is (correctly) empty.
```
```suggestion
        The incoming request is (correctly) empty.
```

todofixthis commented on test/commands/core/get_missing_transactions_test.py:
```diff
@@ -0,0 +1,116 @@
+
+    def test_fail_unexpected_parameters(self):
+        """
+        The incoming response contains unexpected parameters.
```
```suggestion
        The incoming request contains unexpected parameters.
```

todofixthis commented on test/commands/core/get_node_api_configuration_test.py:
```diff
@@ -0,0 +1,95 @@
+
+    def test_pass_empty(self):
+        """
+        The incoming response is (correctly) empty.
```
```suggestion
        The incoming request is (correctly) empty.
```

todofixthis commented on test/commands/core/get_node_api_configuration_test.py:
```diff
@@ -0,0 +1,95 @@
+
+    def test_fail_unexpected_parameters(self):
+        """
+        The incoming response contains unexpected parameters.
```
```suggestion
        The incoming request contains unexpected parameters.
```

todofixthis commented on test/commands/core/get_balances_test.py:
```diff
@@ -14,290 +14,345 @@
+
+            'threshold': 80,
+
+            'tips': [self.trytes3, ],
```
```suggestion
            'tips': [self.trytes3],
```

todofixthis commented on test/commands/core/get_balances_test.py:
```diff
@@ -14,290 +14,345 @@
+            'balances':       ['114544444', '0', '8175737'],
+            'duration':       42,
+            'milestoneIndex': 128,
+            'references':     [self.trytes1, ]
```
```suggestion
            'references':     [self.trytes1]
```

todofixthis commented on test/commands/core/get_balances_test.py:
```diff
@@ -14,290 +14,345 @@
+                'balances':       [114544444, 0, 8175737],
+                'duration':       42,
+                'milestoneIndex': 128,
+                'references':     [self.trytes1, ]
```
```suggestion
                'references':     [self.trytes1]
```

todofixthis commented on test/commands/core/were_addresses_spent_from_test.py:
```diff
@@ -9,167 +9,208 @@
+
+    def test_multiple_addresses(self):
+        """
+        Typical ``wereAddressesSpentFrom`` response for a multiple address
```
```suggestion
        Typical ``wereAddressesSpentFrom`` response for multiple addresses.
```

lzpap commented:
> Minor changes requested, but generally looks really good, thanks @pdecol!
> 
> @lzpap we need to come to a decision about whether/how to validate responses coming from the IRI node (survey app developer community and see if there's a clear preference one way or the other, maybe?).

Hey @todofixthis I will start with coordinating with the other libs to see if they validate requests/responses. I am generally for filters in both directions. Then you could handle exceptions and also it might be easier to reveal bugs with future IRI API changes.

lzpap commented:
> > Minor changes requested, but generally looks really good, thanks @pdecol!
> > @lzpap we need to come to a decision about whether/how to validate responses coming from the IRI node (survey app developer community and see if there's a clear preference one way or the other, maybe?).
> 
> Hey @todofixthis I will start with coordinating with the other libs to see if they validate requests/responses. I am generally for filters in both directions. Then you could handle exceptions and also it might be easier to reveal bugs with future IRI API changes.

@todofixthis I talked with the devs of the other libraries, and the conclusion is that responses are not filtered/validated/checked in general, except bundle and transactions hashes. I think it would be nice to make PyOTA consistent with this behavior.

pdecol commented:
Thanks for the review! I implemented the requested changes and reverted the response filters for the fields that we decided not to validate.

todofixthis commented on iota/commands/core/get_balances.py:
```diff
@@ -44,19 +45,48 @@ def __init__(self):
+                    f.Required |
+                    Trytes(TransactionHash) |
+                    f.Unicode(encoding='ascii', normalize=False),
+                ),
```
❔ Could we remove `f.Required` from the response filter?  Match behaviour of other libs, so that node responses don't generate exceptions.
❔ The `f.Unicode` filter will convert `TransactionHash` values into strings.  This is desirable for request filters (so that the string values can be converted to JSON to send to the node), but probably counter-productive for response filters (the client would likely prefer to work with `TransactionHash` objects rather than strings).  Could we remove it from the response filter?

Apologies for not picking up on this during the first review.

pdecol commented:
No worries. Thanks for spotting this mistake. I removed `f.Required` and `f.Unicode` and I added `f.ByteString` so that it is consistent with the response filters in findTransaction, getTrytes, getTips and getMissingTransactions.


One more thing that I would like to ask you. Can you change this URL here in GitHub? (https://iota.readme.io/ -> https://docs.iota.org/) I think this should be the very last URL that needs to be fixed.

![image](https://user-images.githubusercontent.com/15176378/65830911-2f799a80-e2b4-11e9-940e-37aa460f4afa.png)

todofixthis commented:
LGTM Thanks @pdecol !!

lzpap commented:
Thanks @pdecol !
